cloudinit/distros/opensuse.py

```diff
@@ -66,15 +74,72 @@ def install_packages(self, pkglist):
         )
 
     def package_command(self, command, args=None, pkgs=None):
+        pckg_op = command
         if pkgs is None:
             pkgs = []
 
+        if self.update_method is None:
+            result = util.get_mount_info("/")
+            fs_type = ""
+            if result:
+                (devpth, fs_type, mount_point) = result
+                # Check if the file system is read only
+                mounts = util.load_file("/proc/mounts").split("\n")
+                for mount in mounts:
+                    if mount.startswith(devpth):
+                        mount_info = mount.split()
+                        if mount_info[1] != mount_point:
+                            continue
+                        if mount_info[3].startswith("ro"):
+                            self.read_only_root = True
+                        else:
+                            self.read_only_root = False
+                        break
+                if fs_type.lower() == "btrfs" and os.path.exists(
+                    "/usr/sbin/transactional-update"
+                ):
+                    self.update_method = "transactional"
+                else:
+                    self.update_method = "zypper"
+            else:
+                LOG.info(
+                    "Could not determine filesystem type of '/' using zypper"
+                )
+                self.update_method = "zypper"
+
+        if self.read_only_root and not self.update_method == "transactional":
+            LOG.error(
+                "Package operation requested but read only root "
+                "without btrfs and transactional-updata"
+            )
+            return
+
         # No user interaction possible, enable non-interactive mode
-        cmd = ["zypper", "--non-interactive"]
+        if self.update_method == "zypper":
+            cmd = ["zypper", "--non-interactive"]
+        else:
+            cmd = [
+                "transactional-update",
+                "--non-interactive",
+                "--drop-if-no-change",
+                "pkg",
+            ]
 
         # Command is the operation, such as install
-        if command == "upgrade":
+        if pckg_op == "upgrade":
             command = "update"
+        if self.update_method == "transactional" and not pkgs:
```

Comment:
Nit: It might be a little cleaner to move this into the else above (with a subcondition checking for pkgs). Up to you.

Reply:
This has to be here as the command could be "update" or "upgrade".